### PR TITLE
Introduce implicit association identity provider configuration

### PR DESCRIPTION
**Pull request by Yasasr1**
### Proposed changes in this pull request

With this PR, a new identity provider configuration will be introduced that can be used to configure the implicit association behaviour during federated login. 

```
"implicitAssociation": {
        "isEnabled": true,
        "lookupAttribute": [
            "username",
            "mobile"
        ]
}
```
### Related issues
- https://github.com/wso2-enterprise/asgardeo-product/issues/20046

### CODE REVIEW

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 1]
```diff
@@ -2391,6 +2393,29 @@ private List<IdentityProviderProperty> filterIdentityProperties(IdentityProvider
...
+                    federatedAssociationConfig.setLookupAttributes(attributes);
+                    break;
+            }
+        });
+
+        // For newly created idps default values will be added. However, these will be null for previously
+        // created idps and hence we need to set the default values.
+        if (federatedAssociationConfig.getLookupAttributes() == null) {
+            federatedAssociationConfig.setLookupAttributes(
+                    new String[]{IdPManagementConstants.DEFAULT_LOOKUP_ATTRIBUTE});
```
Comment (pulasthi7) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:2414]: do we need this actually? for old IDPs this config will be disabled anyway.
Also when you enable this config, one of the lookup attributes should be made mandatory, or else the update request should be failed

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 2]
```diff
@@ -3080,6 +3111,14 @@ private List<IdentityProviderProperty> getCombinedProperties(JustInTimeProvision
         associateLocalUser.setName(IdPManagementConstants.ASSOCIATE_LOCAL_USER_ENABLED);
         associateLocalUser.setValue("false");
 
+        IdentityProviderProperty federatedAssociationProperty = new IdentityProviderProperty();
+        federatedAssociationProperty.setName(IdPManagementConstants.FEDERATED_ASSOCIATION_ENABLED);
+        federatedAssociationProperty.setValue("false");
```
Comment (pulasthi7) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:3116]: take this to a constant as well

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 3]
```diff
@@ -3080,6 +3111,14 @@ private List<IdentityProviderProperty> getCombinedProperties(JustInTimeProvision
...
+
+        IdentityProviderProperty lookupAttribute = new IdentityProviderProperty();
+        lookupAttribute.setName(IdPManagementConstants.LOOKUP_ATTRIBUTES);
+        lookupAttribute.setValue(IdPManagementConstants.DEFAULT_LOOKUP_ATTRIBUTE);
```
Comment (pulasthi7) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:3120]: As mentioned above, this seems not necessary

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java (hunk 1 above)
Comment (Yasasr1) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:2414]: I added this to provide a consistent behaviour for old and new idps. So we can enable later on in old idps as well

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java [hunk 4]
```diff
@@ -2391,6 +2393,29 @@ private List<IdentityProviderProperty> filterIdentityProperties(IdentityProvider
...
         }
+
+        identityProviderProperties.forEach(identityProviderProperty -> {
+            switch (identityProviderProperty.getName()) {
+                case IdPManagementConstants.FEDERATED_ASSOCIATION_ENABLED:
+                    federatedAssociationConfig
+                            .setEnabled(Boolean.parseBoolean(identityProviderProperty.getValue()));
+                    break;
+                case IdPManagementConstants.LOOKUP_ATTRIBUTES:
+                    String[] attributes = identityProviderProperty.getValue().split(",");
```
Comment (kayathiri4) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java:2404]: Is there any chance for `identityProviderProperty.getValue()` to be null? If so, shall we add a null check here?

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java (hunk 1 above)
Comment (Yasasr1) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java]: Removed with [3c20442](https://github.com/wso2/carbon-identity-framework/pull/5017/commits/3c20442145ac5549fb011ce7fae823d2970cfbe9)

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java (hunk 2 above)
Comment (Yasasr1) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java]: Addressed with [3c20442](https://github.com/wso2/carbon-identity-framework/pull/5017/commits/3c20442145ac5549fb011ce7fae823d2970cfbe9)

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java (hunk 3 above)
Comment (Yasasr1) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java]: Addressed with [3c20442](https://github.com/wso2/carbon-identity-framework/pull/5017/commits/3c20442145ac5549fb011ce7fae823d2970cfbe9)

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java (hunk 4 above)
Comment (Yasasr1) [reply components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/dao/IdPManagementDAO.java]: Addressed with [3c20442](https://github.com/wso2/carbon-identity-framework/pull/5017/commits/3c20442145ac5549fb011ce7fae823d2970cfbe9)

**Review by kayathiri4 — APPROVED** [review]

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPManagementConstants.java [hunk 5]
```diff
@@ -101,6 +101,11 @@ public class IdPManagementConstants {
     public static final String DEFAULT_SYNC_ATTRIBUTE = "OVERRIDE_ALL";
     public static final String PRESERVE_LOCAL_ATTRIBUTE_SYNC = "PRESERVE_LOCAL";
     public static final String PRESERVE_LOCALLY_ADDED_CLAIMS = "JITProvisioning.PreserveLocallyAddedClaims";
+    public static final String FEDERATED_ASSOCIATION_ENABLED = "FEDERATED_ASSOCIATION_ENABLED";
+    public static final String LOOKUP_ATTRIBUTES = "LOOKUP_ATTRIBUTES";
+    public static final String DEFAULT_LOOKUP_ATTRIBUTE = "email";
+
+    public static final String FALSE = "false";
```
Comment (pulasthi7) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/util/IdPManagementConstants.java:108]:
```suggestion
    public static final String FEDERATED_ASSOCIATION_ENABLED_DEFAULT_VALUE = "false";
```

**Review by pulasthi7 — APPROVED** [review]

**Review by pulasthi7 — DISMISSED** [review]

**Comment by Yasasr1** [comment]
PR builder failed due to following test failing:
JDBCUserStoreAddingTestCase.testAddJDBCUserStore

Tested and verified the test case locally and it is passing
Screenshot 2023-10-18 at 8 31 36 PM

**Review by pulasthi7 — APPROVED** [review]

**Comment by Yasasr1** [comment]
Merging as per above comment

**Pull request merged**